### PR TITLE
[2.0.x] HAL for ESP32

### PR DESCRIPTION
**Pull request by simon-jouet**
Basic HAL for ESP32.

Comments are more than welcome. I will soon push to my repositories the ESP32 project (so you can compile this as an esp-idf project instead of an arduino project) as well as the pinout and wiring i've been using.

Few things must be done before even thinking of merging:

- [ ] Remove Wire and Wire.begin() from the setup
- [x] Remove the commented out sensitive pins in the main. Problem is I use the MSB of the byte to decide if the pin is on the GPIO expander or not and the function converts to int8_t.
- [x] Remove the commented out analogWrite in M42.cpp, like I mentioned in the Marlin 2 thread, I think the analogWrite() should be abstracted away in the HAL to use PWM/DAC ... where suitable
- [ ] Figure out if the GPIO expander should be supported this way. It limits the scope to the HAL but also ties the microcontroller to a specific GPIO expander and doesn't allow for other microcontroller to share the GPIO expander code ...
- [ ] Remove my own settings from the MR (configuration and configuration_adv)

### CODE REVIEW

**Comment by simon-jouet** [comment]
Progress, endstops and thermistors are now working. I had to change the resistors on the ramps from 4.7k to 10k in order to measure low temperatures, but that's a detail for now. Next will be to test if the PWM of the mosfets works fine, if it does then it will be time for a test print :)

**Comment by thinkyhead** [comment]
A build error needs to be resolved.
```
Marlin/src/HAL/HAL_ESP32/fastio_ESP32.cpp: In function 'void set_i2c_pin_mode(uint8_t, uint8_t)':
Marlin/src/HAL/HAL_ESP32/fastio_ESP32.cpp:49:63: error: 'INPUT' was not declared in this scope
set_i2c_bit(GET_REGISTER(pin, IODIRA), GET_PIN(pin), mode & INPUT ? 1 : 0);
```

File: Marlin/src/HAL/HAL_ESP32/fastio_ESP32.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+void set_i2c_bit(uint8_t reg, uint8_t bit, uint8_t val) {
+  uint8_t r = get_i2c_register(reg);
+  r = val ? (r | (1 << bit)) : (r & ~(1 << bit));
+  set_i2c_register(reg, r);
+}
+
+void set_i2c_pin_mode(uint8_t pin, uint8_t mode) {
+  set_i2c_bit(GET_REGISTER(pin, IODIRA), GET_PIN(pin), mode & INPUT ? 1 : 0);
```
Comment (thinkyhead) [inline Marlin/src/HAL/HAL_ESP32/fastio_ESP32.cpp:49]: Note that in `Arduino.h`, `INPUT` is defined as 0. So `mode & INPUT` will always result in 0. I'm guessing that `INPUT` here should be something else. If I can figure it out I'll make the edit.

Comment (simon-jouet) [reply Marlin/src/HAL/HAL_ESP32/fastio_ESP32.cpp:49]: Hi @thinkyhead, thanks for the feedbacks, I  just double checked and in the ESP32 definition INPUT is 1 (https://github.com/espressif/arduino-esp32/blob/master/cores/esp32/esp32-hal-gpio.h#L33)

Comment (thinkyhead) [reply Marlin/src/HAL/HAL_ESP32/fastio_ESP32.cpp:49]: Is that compatible with the `pinMode(pin, INPUT)` function?

Comment (simon-jouet) [reply Marlin/src/HAL/HAL_ESP32/fastio_ESP32.cpp:49]: I guess you mean the fact that Input is not 0? Yes it does work fine, if you look at the HAL I used pinMode for SET_INPUT/SET_OUTPUT for the esp32 pins.

The SET_INPUT/SET_OUTPUT functions will use `set_i2c_pin_mode` if the pin is on the GPIO expander and use `pinMode(pin, INPUT)` if it's one of the ESP pin.

**Comment by luc-github** [comment]
@simon-jouet it is interresting 
Is there already a board designed with ESP32 MCU with RAMPS or equivalent ?
or do you just wire a ramps to a standard ESP32 board ?

**Comment by simon-jouet** [comment]
@luc-github Thanks; right now I'm using an ESP32-T (just look on aliexpress) devboard which is quite basic, a MCP23017 GPIO expander in PDIP and all of that on a protoboard. It's a bit of a mess with all the wires but I got most of it working.

I designed a prototype PCB which I ordered last week so hopefully I will receive it soon. (and hopefully I didn't mess up something too bad in the design). I will upload the current Kicad design and PCB to my github at some point if anybody is interested.

Right now i'm pretty much at the moment of printing but I'm having an issue with sending the Gcode, when I send them one by one it's fine, but when I start a print and the commands are sent sequentially by Pronterface the printer crashes. Anyway I've been trying to get the SDcard to work to check if it's an issue with the serial but didn't had much time.

(sneak peek of the photo sent by the PCB house)
image

**Comment by luc-github** [comment]
Very nice project ^_^
The ESP32 Serial buffer is 256B by default so you may overflow it if ESP32 is too slow to handle each Gcode incoming command 
Did you tried to increase rx buffer to 1KB  ?

**Comment by simon-jouet** [comment]
I haven't, thanks for the input I will give it a try.

Btw, once the boards arrive (and I haven't messed up) I might populate a few and send them around for devs to play around and hopefully help with the Marlin ESP32 HAL. Not too sure how long that will take and how much that will cost to ship things around but that's what I have in mind for now.

**Comment by AnHardt** [comment]
As far as i know `Pronterface ` is still using pure pingpong protocol. So it can't be a matter of speed. Buffer sizes should be larger than the longest command. But if the default for the serial is 256 - thats pretty save. The default for MAX_CMD_SIZE, what is the default for the line buffers, is only 96. Up to now i have not heard about problems with that.

**Comment by davidelang** [comment]
On Wed, 25 Oct 2017, Luc wrote:

> @simon-jouet it is interresting
> Is there already a board designed with ESP32 MCU with RAMPS or equivalent ?
> or do you just wire a ramps to a standard ESP32 board ?

I have a board that is a standard arduino (not mega) pinout, but with an ESP32 
driving it.

such things exist and are going to show up more frequently going forward.

David Lang

**Comment by luc-github** [comment]
@davidelang what is the board you have/describe ? I only know new arduino cinque which has esp32 but drived by stm32

**Comment by thinkyhead** [comment]
I've attempted to resolve conflicts so please double-check that the resolutions are sound.

**Comment by simon-jouet** [comment]
Thanks Scott for the merge, I will have a look at that over the weekend

**Comment by psavva** [comment]
I'll be happy to test your board if you send it over to Cyprus.

**Comment by simon-jouet** [comment]
@luc-github board arrived and I just finished putting one together :)

Board is okay enough to work but has a few issues, I will update the board soon. not too sure if I should bother assembling more of this version. 

Existing issues are mostly from stupid errors that I should have checked beforehand:
- SOIC footprint for the GPIO expander is too slim so had to deadbug half the pins
- the pins of my transistors aren't the same as the one I have in stock, might change to more common ones
- missed a label on the CP2102

and few improvements like adding a few leds for status, at least if the power is properly delivered

(terrible picture)
image

**Comment by luc-github** [comment]
Nice ^_^
So it will accept a Ramps,  right ? but I see buttons (reset and boot?) will be under ramps board and so not accessible - would be more accessible if on side and perpendicular no ?

Did you do some performance test already ?
Did you solved the serial issue ?

**Comment by simon-jouet** [comment]
It does accept a ramps board on top

The buttons are mostly to test things out. I guess they should be removed in the long term as the DTR and RTS line of the cp2102 will automatically assert them for programming.

I haven't had time for any more work or tests unfortunately... Now that this board is soldered and working I will keep going :).

**Comment by luc-github** [comment]
Nice work - keep going - can't wait to see next  update ^_^

**Comment by simon-jouet** [comment]
Got the SDCard to work (after quite a bit of hacking around) and I figured two things out
- What I thought might have been a serial problem isn't, same thing happened over the SDcard, the printer started printing and just crashed. Not too sure how to debug that, I will have to keep looking (i'm suspecting something with the temp isr)
- I currently use the HSPI pins of the ESP32, the idea was to expose them so we can use a SDcard in normal use but we can also use the JTAG debug if necessary. The issue is that most SDcards have internal pullups and it messes up the GPIO 12 at boot time sometimes, preventing boot when the SDcard is in the socket ... might have to shuffle the pins around to use VSPI instead.

@luc-github If you are interested in helping out with this, I can send a board your way, it's far from perfect but it's enough to get the HAL in place :)

**Comment by luc-github** [comment]
Great to see progress - sure if I can help I would be happy to - will PM you later

For PM you can use gitter - most of github user have it as it use github credential

**Comment by tedder** [comment]
I was sent over here from #7076.  How's this going? Does the shield work? How can I help?

**Comment by simon-jouet** [comment]
Hi @tedder,

Have a look at #6524 it's more up-to-date than this thread. We are still working on it, and there is progress, @luc-github and myself now have a board to do some experiment but I had little time to spend on this. I will try to go back to it in the new year.

**Comment by tedder** [comment]
Aha, thanks. Forgot about that (already!). I can always get a couple boards from oshpark if it's helpful.

**Comment by thinkyhead** [comment]
Rebased. Hopefully got the merges correct.

---
It will always be best to use `git rebase` and not `git merge` when maintaining a branch for a long time. The `git merge` command (or "sync" button in Github Desktop) causes new commits to pile on top of your changes, whereas `git rebase` moves your changes after all commits at this point in time.

When `git rebase` is done faithfully on a near-daily basis (along with `git rebase -i`) the working branch remains fresh and up-to-date. With `git merge` (or just letting a branch get stale) the branch becomes more and more of a minefield and more likely to accrue conflict resolution errors.

**Comment by simon-jouet** [comment]
I've cleaned everything, removed the old esp32 branch, rebased the bugfix-2.0.x branch and brought the changes up to date. I will keep working on this and open a MR once it's in a usable state.

@thinkyhead Sorry for the previous fiddling with the merge vs rebase, I will keep using rebase from now on.

**Pull request closed without merging**